### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by Leedukas**
feat(PP-12): added error handling and retries for game data fetching

### CODE REVIEW

File: middleware.js [hunk 1]
```diff
@@ -12,13 +14,33 @@ module.exports.isLoggedIn = (req, res, next) => {
...
+  while (retries <= maxRetries && !success) {
+    try {
+      const randGameId = Math.floor(Math.random() * 500000 + 1);
+      response = await axios.get(
+        `https://api.rawg.io/api/games/${randGameId}?key=${process.env.RAWG_KEY}`
+      );
+      success = true;
+      break;
+    } catch (error) {
+      console.log(error.response.status);
```
Comment (elnino97) [inline middleware.js:31]: For the future let's implement showing the error in console only if in development

File: package.json [hunk 2]
```diff
@@ -11,6 +11,7 @@
   "license": "ISC",
   "dependencies": {
     "axios": "^1.3.4",
+    "axios-retry": "^3.4.0",
```
Comment (elnino97) [inline package.json:14]: This package is unused. You can remove it typing `npm uninstall axios-retry`

File: routes/main.js [hunk 3]
```diff
@@ -10,6 +10,11 @@ router.route(ROUTES.index).get((req, res) => {
   res.render("home", { extraStyles });
 });
 
+router.route(ROUTES.error).get((req, res) => {
+  const extraStyles = '<link rel="stylesheet" href="/stylesheets/cards.css" />';
```
Comment (elnino97) [inline routes/main.js:14]: I would recommend to create a file, maybe `styles.js` or something that exports these constants.

File: views/error.ejs [hunk 4]
```diff
@@ -1,10 +1,9 @@
 <% layout('layouts/boilerplate')%>
-<div class="row">
-  <div class="col-6 offset-3">
-    <div class="alert alert-danger" role="alert">
-      <h4 class="alert-heading"><%=err.message%></h4>
+<div class="card">
+  <div class="card-box">
+    <div class="field-box">
```
Comment (elnino97) [inline views/error.ejs:4]: `field-box` - seems that it was meant for the form elements? Besides there is not a class with the name of `field-box`.

File: middleware.js [hunk 5]
```diff
@@ -12,13 +14,33 @@ module.exports.isLoggedIn = (req, res, next) => {
...
+  let retries = 1;
+  let success = false;
+  const maxRetries = 5;
 
-  next();
+  while (retries <= maxRetries && !success) {
+    try {
+      const randGameId = Math.floor(Math.random() * 500000 + 1);
+      response = await axios.get(
+        `https://api.rawg.io/api/games/${randGameId}?key=${process.env.RAWG_KEY}`
```
Comment (elnino97) [inline middleware.js:26]: I would want to see the API url to declared in a constant

File: middleware.js (hunk 1 above)
Comment (Leedukas) [reply middleware.js:31]: added a .env variable together with conditionals for some console.logs like this one

File: package.json (hunk 2 above)
Comment (Leedukas) [reply package.json:14]: this thing is gone now

File: views/error.ejs (hunk 4 above)
Comment (Leedukas) [reply views/error.ejs:4]: field-box'es are no more

File: middleware.js (hunk 5 above)
Comment (Leedukas) [reply middleware.js:26]: declared it in the file, might move it elsewhere later depending on how the code will develop

File: controllers/users.js [hunk 6]
```diff
@@ -2,7 +2,8 @@ const User = require("../models/user");
 const { getPixelatedImage } = require("../middleware");
 const { ROUTES } = require("./routes");
 
-const extraStyles = '<link rel="stylesheet" href="/stylesheets/cards.css" />';
+const style = require("../public/javascripts/extraStyles.js");
+const extraStyles = style.cardStyle;
```
Comment (elnino97) [inline controllers/users.js:6]: You can destructure the key variable pair from the import above so no need to declare this constant:

```const { cardStyle } = require("../public/javascripts/extraStyles.js");```

File: middleware.js [hunk 7]
```diff
@@ -2,6 +2,7 @@ const Jimp = require("jimp");
 const axios = require("axios");
 
 const { ROUTES } = require("./controllers/routes");
+const apiUrl = "https://api.rawg.io/api/";
```
Comment (elnino97) [inline middleware.js:5]: Let's separate imports from variable declarations with an empty space

File: routes/main.js [hunk 8]
```diff
@@ -4,14 +4,15 @@ const router = express.Router();
 const main = require("../controllers/main");
 const { fetchRandomGameData } = require("../middleware");
 const { ROUTES } = require("../controllers/routes");
+const style = require("../public/javascripts/extraStyles");
 
 router.route(ROUTES.index).get((req, res) => {
-  const extraStyles = '<link rel="stylesheet" href="/stylesheets/main.css" />';
+  const extraStyles = style.mainStyle;
```
Comment (elnino97) [inline routes/main.js:10]: Same as this comment https://github.com/Leedukas/pixel-thing/pull/8/commits/91dce2cba82f659feadec6c7554af81cf9b7f036#r1160728518

File: routes/main.js [hunk 9]
```diff
@@ -4,14 +4,15 @@ const router = express.Router();
...
   res.render("home", { extraStyles });
 });
 
 router.route(ROUTES.error).get((req, res) => {
-  const extraStyles = '<link rel="stylesheet" href="/stylesheets/cards.css" />';
+  const extraStyles = style.cardStyle;
```
Comment (elnino97) [inline routes/main.js:15]: Same here: https://github.com/Leedukas/pixel-thing/pull/8/commits/91dce2cba82f659feadec6c7554af81cf9b7f036#r1160728518

File: controllers/main.js [hunk 10]
```diff
@@ -1,7 +1,9 @@
 const { getPixelatedImage } = require("../middleware");
 
+const style = require("../public/javascripts/extraStyles.js");
```
Comment (elnino97) [inline controllers/main.js:3]: Refer to this comment: https://github.com/Leedukas/pixel-thing/pull/8/commits/91dce2cba82f659feadec6c7554af81cf9b7f036#r1160728518

File: controllers/users.js (hunk 6 above)
Comment (Leedukas) [reply controllers/users.js:6]: is it better to destructure in cases where more than one extra style is being used? (main.js)

File: views/main/play.ejs [hunk 11]
```diff
@@ -9,22 +9,20 @@
...
-            name="gamename"
-            autofocus
-            required
-          />
-        </div>
-        <button class="card-button">Next</button>
-      </form>
-    </div>
+    <h5 class="card-title">What form of game is this game?</h5>
+    <form action="/play" method="POST" class="validated-form" novalidate>
```
Comment (elnino97) [inline views/main/play.ejs:13]: Is `validated-form` class still in use?

File: controllers/users.js (hunk 6 above)
Comment (elnino97) [reply controllers/users.js:6]: Yeah, generally it's better, you can descructure whatever you need. `const { cardStyle, mainStyle } = require('../public/javascripts/extraStyles.js"`. 

Can also just import the `styles` object and use it like this: `extraStyles: styles.cardStyle` etc.

File: controllers/main.js [hunk 12]
```diff
@@ -1,9 +1,9 @@
 const { getPixelatedImage } = require("../middleware");
 
-const style = require("../public/javascripts/extraStyles.js");
+const { cardStyle } = require("../public/javascripts/extraStyles.js");
 
 module.exports.renderPlay = async (req, res, next) => {
   const image = await getPixelatedImage(req.gameData.background_image);
-  const extraStyles = style.cardStyle;
+  const extraStyles = cardStyle;
```
Comment (elnino97) [inline controllers/main.js:7]: Ok this looks better already, but I what I meant is that you don't need to declare a constant with the name `extraStyles` to be able to pass it below. 

Take a look at this: 

```res.render("main/play.ejs", { image, extraStyles: cardStyle });```

Can safely remove the `extraStyles` constant you declare in the route!

**Review by elnino97 — APPROVED** [review]
Good work

**Pull request merged**